synapseclient/core/upload/multipart_upload.py

```diff
@@ -3,385 +3,526 @@
-                         % (float(partSize)/MB, int(math.ceil(fileSize / partSize))))
-    return partSize
 
+thread_local = threading.local()
+
+
+class UploadAbortedException(Exception):
+    """Raised when a worker thread detects the upload was
+    aborted and stops further processing."""
+    pass
+
+
+class UploadFailedException(Exception):
+    """Raised when an upload failed. Should be chained to a cause Exception"""
+    pass
+
+
+class UploadAttempt:
+
+    def __init__(
+        self,
+        syn,
+        chunk_fn,
+        dest_file_name: str,
+        file_size: int,
+        part_size: int,
+        md5_hex: str,
+        content_type: str,
+        preview: bool,
+        storage_location_id: str,
+        max_workers: int,
+        force_restart: bool,
+    ):
+        self._syn = syn
+        self._chunk_fn = chunk_fn
+        self._dest_file_name = dest_file_name
+        self._file_size = file_size
+        self._part_size = part_size
+        self._md5_hex = md5_hex
+        self._content_type = content_type
+        self._preview = preview
+        self._storage_location_id = storage_location_id
+        self._max_workers = max_workers
+        self._force_restart = force_restart
+
+        self._lock = threading.Lock()
+        self._aborted = False
+
+        # populated later
+        self._upload_id: str = None
+        self._pre_signed_part_urls: OrderedDict[int, str] = None
+
+    @classmethod
+    def _get_remaining_part_numbers(cls, upload_status):
+        part_numbers = []
+        parts_state = upload_status['partsState']
+
+        # parts are 1-based
+        for i, part_status in enumerate(parts_state, 1):
+            if part_status == '0':
+                part_numbers.append(i)
+
+        return len(parts_state), part_numbers
+
+    @classmethod
+    def _get_thread_session(cls):
+        # get a lazily initialized requests.Session from the thread.
+        # we want to share a requests.Session over the course of a thread
+        # to take advantage of persistent http connection. we put it on a
+        # thread local rather that in the task closure since a connection can
+        # be reused across separate part uploads so no reason to restrict it
+        # per worker task.
+        session = getattr(thread_local, 'session', None)
+        if not session:
+            session = thread_local.session = requests.Session()
+        return session
+
+    def _create_synapse_upload(self):
+        upload_request = {
+            'contentMD5Hex': self._md5_hex,
+            'fileName': self._dest_file_name,
+            'generatePreview': self._preview,
+            'contentType': self._content_type,
+            'partSizeBytes': self._part_size,
+            'fileSizeBytes': self._file_size,
+            'storageLocationId': self._storage_location_id,
+        }
+
+        return self._syn.restPOST(
+            "/file/multipart?forceRestart={}".format(
```

Comment:
we can use f-strings now that we require 3.6+

Reply:
In this case the string that's being inserted is an instance variable that's being lowered, inline in the format so it seems a bit awkward to use an f-string here.